### ISSUE TITLE
Issue with pushing messages to iOS client

### ISSUE THREAD

**Issue by Vasto** [SignalR/SignalR#4337]
Hi. We are developing a chat functionality for our iOS and Android mobile applications. We are using SignalR hub to receive a message from one mobile, and then send it to another. Unfortunately, we have some issues when the communication involves iOS client. After exchanging several messages between two mobiles, callbacks on iOS app are being no longer called by the hub until a new message from that client is sent to the server. This causes overdue messages to appear on that client. 

We enabled SignalR tracing but the logs don't show any errors. Only thing which we were able to figure out from the logs, was the difference between transport methods used by our mobile apps. iOS
app uses Long Polling, while Android is able to use Server Sent Events. We tried to manually switch to other protocols than Long Polling on iOS client, but the connection attempt always ended with a failure. Here are SignalR logs from those attempts:

Attempt to connect from iOS using Server Sent Events 
```SignalR.Transports.TransportHeartBeat Information: 0 : Connection 4f994e29-9279-4fd1-9346-25dac056a26c is New.
SignalR.Transports.ServerSentEventsTransport Verbose: 0 : Sending outgoing message. Connection id: 4f994e29-9279-4fd1-9346-25dac056a26c, transport: ServerSentEventsTransport, message: data: {"C":"d-D7EE374B-k,0|l,0|m,1","S":1,"M":[]}
SignalR.Transports.ServerSentEventsTransport Verbose: 0 : Sending outgoing message. Connection id: 4f994e29-9279-4fd1-9346-25dac056a26c, transport: ServerSentEventsTransport, message: {"I":"0","E":"There was an error invoking Hub method 'livechat.SendMessage'."}
SignalR.Transports.ServerSentEventsTransport Verbose: 0 : DrainWrites(4f994e29-9279-4fd1-9346-25dac056a26c)
SignalR.Transports.ServerSentEventsTransport Information: 0 : CompleteRequest (4f994e29-9279-4fd1-9346-25dac056a26c)
SignalR.Transports.ServerSentEventsTransport Information: 0 : Abort(4f994e29-9279-4fd1-9346-25dac056a26c)
SignalR.Transports.TransportHeartBeat Information: 0 : Removing connection 4f994e29-9279-4fd1-9346-25dac056a26c
SignalR.Transports.ServerSentEventsTransport Information: 0 : End(4f994e29-9279-4fd1-9346-25dac056a26c)
SignalR.Transports.ServerSentEventsTransport Verbose: 0 : DrainWrites(4f994e29-9279-4fd1-9346-25dac056a26c)
SignalR.Transports.ServerSentEventsTransport Information: 0 : CompleteRequest (4f994e29-9279-4fd1-9346-25dac056a26c)
```
Attempt to connect from iOS using Websockets
``` 
SignalR.Transports.TransportHeartBeat Information: 0 : Connection 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d is New.
SignalR.Transports.WebSocketTransport Verbose: 0 : Sending outgoing message. Connection id: 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d, transport: WebSocketTransport, message: {"C":"d-5C8DA49F-B,0|C,0|D,1","S":1,"M":[]}
SignalR.Transports.WebSocketTransport Error: 0 : OnError(84ec35ad-6bc0-4c9f-9535-36ceabd4a77d, System.Net.WebSockets.WebSocketException (0x80070026): Reached the end of the file
   at System.Web.WebSockets.WebSocketPipe.<>c__DisplayClass9_0.<ReadFragmentAsync>b__0(Int32 hrError, Int32 cbIO, Boolean fUtf8Encoded, Boolean fFinalFragment, Boolean fClose)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.WebSockets.AspNetWebSocket.<DoWork>d__45`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.WebSockets.AspNetWebSocket.<>c__DisplayClass36_0.<<ReceiveAsyncImpl>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketMessageReader.<ReadMessageAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.<ProcessWebSocketRequestAsync>d__25.MoveNext())
SignalR.Transports.WebSocketTransport Information: 0 : CloseSocket(84ec35ad-6bc0-4c9f-9535-36ceabd4a77d)
SignalR.Transports.TransportHeartBeat Verbose: 0 : 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d is dead
SignalR.Transports.TransportHeartBeat Verbose: 0 : 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d is dead
SignalR.Transports.TransportHeartBeat Verbose: 0 : 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d is dead
SignalR.Transports.TransportHeartBeat Verbose: 0 : 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d is dead
SignalR.Transports.TransportHeartBeat Verbose: 0 : 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d is dead
SignalR.Transports.TransportHeartBeat Verbose: 0 : 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d is dead
SignalR.Transports.TransportHeartBeat Verbose: 0 : 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d is dead
SignalR.Transports.TransportHeartBeat Information: 0 : Removing connection 84ec35ad-6bc0-4c9f-9535-36ceabd4a77d
SignalR.Transports.WebSocketTransport Information: 0 : Abort(84ec35ad-6bc0-4c9f-9535-36ceabd4a77d)
SignalR.Transports.WebSocketTransport Information: 0 : End(84ec35ad-6bc0-4c9f-9535-36ceabd4a77d)
SignalR.Transports.WebSocketTransport Verbose: 0 : Cancel(84ec35ad-6bc0-4c9f-9535-36ceabd4a77d)
SignalR.Transports.WebSocketTransport Verbose: 0 : DrainWrites(84ec35ad-6bc0-4c9f-9535-36ceabd4a77d)
SignalR.Transports.WebSocketTransport Information: 0 : CompleteRequest (84ec35ad-6bc0-4c9f-9535-36ceabd4a77d)
SignalR.Transports.TransportHeartBeat Information: 0 : Dispose(). Closing all connections
```

We are using SignalR v.2.2.2. but we were also trying v.2.4.0.

iOS mobile uses SwiftR client for SignalR:
https://github.com/adamhartford/SwiftR

Hub is part of .NET 4.6 Web API application which is hosted on IIS 10.

We’ve done our best to research what the possible problems are, but we’re coming back empty handed and quickly running out of solutions. Can anybody help us out?

**Comment by halter73** [comment]
Are you only seeing these issues with the SwiftR client? If so, please file an issue with them.

**Comment by jeffreymendez1993** [comment]
@Vasto  https://realm.io/products/realm-database/ may be a better solution =)

**Comment by Vasto** [comment]
It seems we have only issues with SwiftR. But I have one more question, which arose when we were trying to figure out why our iOS client is not able negotiate anything else than Long Polling.

We noticed that when we try to establish a Websocket connection, and the client sends a Authorization header with a JWT token, we everytime get the following error:

`Unrecognized user identity.  The user identity cannot change during an active SignalR connection`

When we remove the header we are able to establish connection successfully, but then we are not able to invoke our hub method which requires authorization.

`SignalR.Transports.WebSocketTransport Verbose: 0 : Sending outgoing message. Connection id: 337c0b4b-03b9-42ca-bbbb-eec7551e42e3, transport: WebSocketTransport, message: {"I":"0","E":"Caller is not authorized to invoke the SendMessage method on xxx."}`

Any suggestion how we could resolve this issue?

**Comment by halter73** [comment]
@Vasto is the same Authorization set in the /negotiate requests preceding the /connect request? If not, that could explain the behavior you're seeing.

**Comment by Vasto** [comment]
Ok, we finally were able to solve the issue.

SwiftR is basically a wrapper for JavaScript SignalR client, and according to this [thread](https://github.com/aspnet/SignalR/issues/130) WebSockets JavaScript client doesn't support custom headers. I realize the linked thread is about SignalR for .NET Core, but it seems that similar limitation applies also here. 

After we changed the method of sending token from headers to query string and modified backend with a middleware which takes token from a query and adds it as Authorization header, we finally were able to establish Server Sent Events and Websocket connections from iOS.

Moving from Long Polling to Websockets or Server Sent Events, also resolved our issue with callbacks being not called on the iOS client after sending several messages.
